File: extension/src/filemanager.ts
import * as fs from 'fs/promises';
import * as path from 'path';
import * as os from 'os';
import * as vault from './vault';
import * as recovery from './recovery';
import * as envparser from './envparser';
import * as detector from './detector';
import * as sandbox from './sandbox';
import { MARKER_FORMAT_VERSION } from './version';

export interface CloakMarker {
    version: number;
    protected: string[];
    projectHash: string;
    createdAt: string;
}

export interface ProtectResult {
    secretCount: number;
    alreadyProtected: boolean;
}

// ─────────────────────────────────────────────────────────────────────────────
// Platform paths
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Returns the platform-appropriate vaults directory, creating it if needed.
 * - macOS: ~/Library/Application Support/cloak/vaults/
 * - Linux: ~/.config/cloak/vaults/
 * - Windows: %APPDATA%\cloak\vaults\
 */
async function vaultsDir(): Promise<string> {
    let base: string;
    if (process.platform === 'darwin') {
        base = path.join(os.homedir(), 'Library', 'Application Support', 'cloak', 'vaults');
    } else if (process.platform === 'win32') {
        const appdata = process.env['APPDATA'];
        if (!appdata) throw new Error('APPDATA environment variable not set');
        base = path.join(appdata, 'cloak', 'vaults');
    } else {
        // Linux / other — prefer XDG_CONFIG_HOME
        const xdg = process.env['XDG_CONFIG_HOME'];
        base = xdg
            ? path.join(xdg, 'cloak', 'vaults')
            : path.join(os.homedir(), '.config', 'cloak', 'vaults');
    }
    await fs.mkdir(base, { recursive: true });
    return base;
}

/**
 * Returns the vault file path for `relPath` within the given project root.
 * Format: <vaultsDir>/<vaultId>.vault — one vault per protected file so that
 * multiple `.env` files in a project never collide (see vault.vaultId).
 */
export async function vaultFilePath(projectRoot: string, relPath: string): Promise<string> {
    const hash = vault.projectHash(projectRoot);
    const id = vault.vaultId(hash, relPath);
    const dir = await vaultsDir();
    return path.join(dir, `${id}.vault`);
}

/**
 * Returns the recovery file path for the given project root.
 * Format: <vaultsDir>/<projectHash>.recovery
 */
export async function recoveryFilePath(projectRoot: string): Promise<string> {
    const hash = vault.projectHash(projectRoot);
    const dir = await vaultsDir();
    return path.join(dir, `${hash}.recovery`);
}

/**
 * Returns the auth file path for the given project root.
 * Format: <vaultsDir>/<projectHash>.auth — must match the CLI (auth.rs).
 */
export async function authFilePath(projectRoot: string): Promise<string> {
    const hash = vault.projectHash(projectRoot);
    const dir = await vaultsDir();
    return path.join(dir, `${hash}.auth`);
}

// ─────────────────────────────────────────────────────────────────────────────
// Recursive .env discovery (mirrors the CLI's `scan_env_files`)
// ─────────────────────────────────────────────────────────────────────────────

/** Candidate env file names scanned in each directory. Matches the CLI. */
const ENV_FILE_NAMES = [
    '.env',
    '.env.local',
    '.env.development',
    '.env.production',
    '.env.staging',
    '.env.test',
];

/** Directory names skipped during the scan (dot-directories are skipped separately). */
const IGNORE_DIRS = new Set(['node_modules', 'dist', 'build', 'target', 'vendor', 'coverage']);

/** Maximum directory depth to descend (0 = project root only). */
const MAX_SCAN_DEPTH = 5;

/**
 * Recursively scan `rootDir` (subdirectories up to {@link MAX_SCAN_DEPTH}) for
 * candidate env files that contain at least one detected secret.
 *
 * Returns `{ relPath, secretCount }` for each, with forward-slash relative paths
 * sorted deterministically. `node_modules`, build output, and dot-directories are
 * skipped, and symlinked directories are not followed — same rules as `cloak init`.
 */
export async function findEnvFilesWithSecrets(
    rootDir: string,
): Promise<Array<{ relPath: string; secretCount: number }>> {
    const results: Array<{ relPath: string; secretCount: number }> = [];

    async function scan(dir: string, depth: number): Promise<void> {
        // Candidate env files in this directory.
        for (const name of ENV_FILE_NAMES) {
            const full = path.join(dir, name);
            let content: string;
            try {
                content = await fs.readFile(full, 'utf8');
            } catch {
                continue; // not a file / unreadable
            }
            const secretCount = envparser.parse(content).filter(
                line => line.type === 'assignment' && detector.detect(line.key, line.value).isSecret,
            ).length;
            if (secretCount > 0) {
                results.push({
                    relPath: path.relative(rootDir, full).replace(/\\/g, '/'),
                    secretCount,
                });
            }
        }

        if (depth >= MAX_SCAN_DEPTH) return;

        let entries: import('fs').Dirent[];
        try {
            entries = await fs.readdir(dir, { withFileTypes: true });
        } catch {
            return;
        }
        for (const entry of entries) {
            // isDirectory() is false for symlinks, so symlinked dirs are skipped.
            if (!entry.isDirectory()) continue;
            if (entry.name.startsWith('.') || IGNORE_DIRS.has(entry.name)) continue;
            const child = path.join(dir, entry.name);
            // Don't descend into an independently-protected sub-project — its own
            // .cloak owns those files.
            try {
                await fs.access(path.join(child, '.cloak'));
                continue;
            } catch { /* no marker here — keep scanning */ }
            await scan(child, depth + 1);
        }
    }

    await scan(rootDir, 0);
    results.sort((a, b) => a.relPath.localeCompare(b.relPath));
    return results;
}

// ─────────────────────────────────────────────────────────────────────────────
// Atomic write helpers
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Atomically write bytes to a file.
 * Writes to <path>.tmp, then renames. On Windows removes original first.
 */
async function atomicWriteBytes(filePath: string, data: Buffer): Promise<void> {
    const tmpPath = filePath + '.tmp';
    await fs.writeFile(tmpPath, data);
    if (process.platform === 'win32') {
        try { await fs.unlink(filePath); } catch { /* file may not exist */ }
    }
    await fs.rename(tmpPath, filePath);
}

/**
 * Atomically write a UTF-8 string to a file.
 */
async function atomicWriteStr(filePath: string, text: string): Promise<void> {
    await atomicWriteBytes(filePath, Buffer.from(text, 'utf8'));
}

// ─────────────────────────────────────────────────────────────────────────────
// Timestamp helper
// ─────────────────────────────────────────────────────────────────────────────

function iso8601Now(): string {
    return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

// ─────────────────────────────────────────────────────────────────────────────
// Public API
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Protect a .env file: detect secrets → encrypt vault → write recovery →
 * write sandbox to disk → update .cloak marker.
 */
export async function protectFile(
    projectRoot: string,
    relPath: string,
    key: Buffer,
    recoveryKeyBytes: Buffer | null,
): Promise<ProtectResult> {
    const envPath = path.join(projectRoot, relPath);

    // 1. Read the .env file.
    const content = await fs.readFile(envPath, 'utf8');

    // 2. Parse into EnvLine values.
    const lines = envparser.parse(content);

    // 3. Count secrets.
    const secretCount = lines.filter(line => {
        if (line.type === 'assignment') {
            return detector.detect(line.key, line.value).isSecret;
        }
        return false;
    }).length;

    // 4. If no secrets, skip protection.
    if (secretCount === 0) {
        return { secretCount: 0, alreadyProtected: false };
    }

    // 5. Get project hash.
    const hash = vault.projectHash(projectRoot);

    // Check if already protected (vault already exists).
    const vPath = await vaultFilePath(projectRoot, relPath);
    let alreadyProtected = false;
    try {
        await fs.access(vPath);
        alreadyProtected = true;
    } catch { /* vault doesn't exist */ }

    // 6. Generate sandbox content.
    const sandboxContent = sandbox.sandboxEnv(content, hash);

    // 7. Encrypt original content.
    const vaultBytes = vault.encrypt(content, key);

    // 8. Write vault file atomically.
    await atomicWriteBytes(vPath, vaultBytes);

    // 9. Write the per-project recovery file — only when seeding a fresh key.
    //    When adding a file to an already-protected project the key is reused and
    //    recoveryKeyBytes is null, so the existing recovery file is left untouched
    //    (rewriting it would invalidate the recovery key the user already saved).
    if (recoveryKeyBytes) {
        const recoveryBytes = recovery.createRecoveryFile(key, recoveryKeyBytes);
        const rPath = await recoveryFilePath(projectRoot);
        await atomicWriteBytes(rPath, recoveryBytes);
        if (process.platform !== 'win32') {
            await fs.chmod(rPath, 0o600);
        }
    }

    // 10. Set permissions 600 on the vault file (Unix).
    if (process.platform !== 'win32') {
        await fs.chmod(vPath, 0o600);
    }

    // 12. Write sandbox content to the .env file on disk (atomic).
    await atomicWriteStr(envPath, sandboxContent);

    // 13. Update .cloak marker.
    const marker = await buildOrUpdateMarker(projectRoot, relPath, hash);
    await writeMarker(projectRoot, marker);

    return { secretCount, alreadyProtected };
}

/**
 * Remove protection: decrypt vault → restore original .env → remove vault →
 * update .cloak marker.
 */
export async function unprotectFile(
    projectRoot: string,
    relPath: string,
    key: Buffer,
): Promise<void> {
    // Decrypt vault and get real content.
    const realContent = await readReal(projectRoot, relPath, key);

    // Restore the real .env to disk.
    const envPath = path.join(projectRoot, relPath);
    await atomicWriteStr(envPath, realContent);

    // Remove vault file.
    const vPath = await vaultFilePath(projectRoot, relPath);
    try {
        await fs.unlink(vPath);
    } catch { /* vault may not exist */ }

    // Update marker — remove this file from the protected list.
    const marker = await readMarker(projectRoot);
    if (marker) {
        marker.protected = marker.protected.filter(p => p !== relPath);
        await writeMarker(projectRoot, marker);
    }
}

/**
 * Read the real (decrypted) content of a protected .env file from its vault.
 */
export async function readReal(
    projectRoot: string,
    relPath: string,
    key: Buffer,
): Promise<string> {
    const vPath = await vaultFilePath(projectRoot, relPath);

    // Check for missing vault.
    let vaultExists = false;
    try {
        await fs.access(vPath);
        vaultExists = true;
    } catch { /* vault missing */ }

    if (!vaultExists) {
        const marker = await readMarker(projectRoot);
        if (marker) {
            throw new Error('Vault file missing. Run cloak recover to restore access.');
        }
        throw new Error(`No vault found for ${relPath}. Has this file been protected?`);
    }

    const vaultBytes = await fs.readFile(vPath);

    // Sanity check.
    if (!vault.isVault(vaultBytes)) {
        throw new Error('Vault corrupted. If you have your recovery key, run cloak recover.');
    }

    const plaintext = vault.decrypt(vaultBytes, key);

    // Basic integrity check.
    if (!plaintext.includes('=')) {
        throw new Error(
            'Vault corrupted: decrypted content does not look like a .env file. ' +
            'If you have your recovery key, run cloak recover.'
        );
    }

    return plaintext;
}

/**
 * Save new real content: encrypt to vault and write sandbox version to disk.
 */
export async function saveReal(
    projectRoot: string,
    relPath: string,
    content: string,
    key: Buffer,
): Promise<void> {
    const vPath = await vaultFilePath(projectRoot, relPath);
    const hash = vault.projectHash(projectRoot);
    const envPath = path.join(projectRoot, relPath);

    // Encrypt new content.
    const vaultBytes = vault.encrypt(content, key);

    // Generate sandbox content.
    const sandboxContent = sandbox.sandboxEnv(content, hash);

    // Write vault atomically.
    await atomicWriteBytes(vPath, vaultBytes);

    // Write sandbox to disk atomically.
    await atomicWriteStr(envPath, sandboxContent);
}

/**
 * Walk up from `startDir` to find the nearest directory containing a `.cloak` marker.
 * Stops at `stopAt` (inclusive). Returns null if no marker is found.
 */
export async function findProjectRoot(startDir: string, stopAt: string): Promise<string | null> {
    let dir = startDir;
    while (dir.startsWith(stopAt)) {
        const markerPath = path.join(dir, '.cloak');
        try {
            await fs.access(markerPath);
            return dir;
        } catch { /* not found */ }
        const parent = path.dirname(dir);
        if (parent === dir) break;
        dir = parent;
    }
    return null;
}

/**
 * Recursively find all directories containing `.cloak` markers within `rootDir`.
 * Returns an array of directory paths.
 */
export async function findAllCloakProjects(rootDir: string): Promise<string[]> {
    const results: string[] = [];
    async function scan(dir: string, depth: number): Promise<void> {
        if (depth > 5) return; // limit depth to avoid scanning node_modules etc.
        const markerPath = path.join(dir, '.cloak');
        try {
            await fs.access(markerPath);
            results.push(dir);
        } catch { /* no marker here */ }
        try {
            const entries = await fs.readdir(dir, { withFileTypes: true });
            for (const entry of entries) {
                if (!entry.isDirectory()) continue;
                if (entry.name.startsWith('.') || entry.name === 'node_modules' || entry.name === 'dist' || entry.name === 'build') continue;
                await scan(path.join(dir, entry.name), depth + 1);
            }
        } catch { /* unreadable dir */ }
    }
    await scan(rootDir, 0);
    return results;
}

/**
 * Read the .cloak marker file from <projectRoot>/.cloak.
 * Returns null if the file does not exist.
 */
export async function readMarker(projectRoot: string): Promise<CloakMarker | null> {
    const markerPath = path.join(projectRoot, '.cloak');
    try {
        const content = await fs.readFile(markerPath, 'utf8');
        return JSON.parse(content) as CloakMarker;
    } catch (err: unknown) {
        if (isNodeError(err) && err.code === 'ENOENT') return null;
        throw err;
    }
}

/**
 * Write (or overwrite) the .cloak marker file at <projectRoot>/.cloak.
 * Written atomically.
 */
export async function writeMarker(projectRoot: string, marker: CloakMarker): Promise<void> {
    const markerPath = path.join(projectRoot, '.cloak');
    const json = JSON.stringify(marker, null, 2);
    await atomicWriteStr(markerPath, json);
}

// ─────────────────────────────────────────────────────────────────────────────
// Private helpers
// ─────────────────────────────────────────────────────────────────────────────

async function buildOrUpdateMarker(
    projectRoot: string,
    relPath: string,
    projectHash: string,
): Promise<CloakMarker> {
    const existing = await readMarker(projectRoot);
    const marker: CloakMarker = existing ?? {
        version: MARKER_FORMAT_VERSION,
        protected: [],
        projectHash,
        createdAt: iso8601Now(),
    };

    if (!marker.protected.includes(relPath)) {
        marker.protected.push(relPath);
    }

    return marker;
}

function isNodeError(err: unknown): err is NodeJS.ErrnoException {
    return typeof err === 'object' && err !== null && 'code' in err;
}

File: cli/src/filemanager.rs
//! File manager — orchestration layer for Cloak's protect/unprotect/read/save operations.
//!
//! This module ties together [`vault`], [`recovery`], [`envparser`], [`detector`],
//! [`sandbox`], and [`platform`] into file-level operations. It is the primary
//! entry point for the CLI commands.
//!
//! # Atomic writes
//!
//! All writes use a write-to-`.tmp`-then-rename strategy to avoid partial writes
//! corrupting files. On Unix the rename is atomic; on Windows the original is
//! removed first (best-available, not truly atomic).
//!
//! # Vault storage
//!
//! Vault and recovery files are stored outside the project directory in the
//! platform config dir (e.g. `~/Library/Application Support/cloak/vaults/`).
//! Only the sandbox `.env` is written inside the project directory.

use crate::{detector, envparser, platform, recovery, sandbox, vault, version};
use anyhow::{anyhow, Context, Result};
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};

// ─────────────────────────────────────────────────────────────────────────────
// Public types
// ─────────────────────────────────────────────────────────────────────────────

/// The `.cloak` marker file that records which files are protected.
///
/// Serialized as JSON and written to `<project_root>/.cloak`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CloakMarker {
    /// Marker format version — always [`version::MARKER_FORMAT_VERSION`].
    pub version: u32,
    /// Relative paths of the files currently under Cloak protection.
    pub protected: Vec<String>,
    /// SHA-256-based project identifier (first 16 hex chars of the path hash).
    #[serde(rename = "projectHash")]
    pub project_hash: String,
    /// ISO 8601-style timestamp when protection was first applied.
    #[serde(rename = "createdAt")]
    pub created_at: String,
}

/// Result returned by [`protect_file`].
#[allow(dead_code)]
pub struct ProtectResult {
    /// Number of secret key-value pairs detected in the file.
    pub secret_count: usize,
    /// `true` when the file was already protected (vault already existed).
    pub already_protected: bool,
}

// ─────────────────────────────────────────────────────────────────────────────
// Public API
// ─────────────────────────────────────────────────────────────────────────────

/// Protect a `.env` file: detect secrets → encrypt vault → write recovery →
/// write sandbox to disk → update `.cloak` marker.
///
/// # Arguments
///
/// * `project_root` — absolute path to the project directory.
/// * `rel_path` — path to the `.env` file, relative to `project_root`.
/// * `key` — 32-byte AES-256-GCM encryption key.
/// * `recovery_key_bytes` — raw bytes of the recovery key used to encrypt the
///   recovery file (see [`recovery::create_recovery_file`]).
///
/// # Returns
///
/// A [`ProtectResult`] describing how many secrets were found and whether the
/// file was already protected.
///
/// # Errors
///
/// Returns an error if the `.env` file cannot be read, vault encryption fails,
/// or any file I/O operation fails.
pub fn protect_file(
    project_root: &Path,
    rel_path: &str,
    key: &[u8; 32],
    recovery_key_bytes: &[u8],
) -> Result<ProtectResult> {
    let env_path = project_root.join(rel_path);

    // 1. Read the .env file.
    let content = std::fs::read_to_string(&env_path)
        .with_context(|| format!("Failed to read {}", env_path.display()))?;

    // 2. Parse into EnvLine values.
    let lines = envparser::parse(&content);

    // 3. Count secrets.
    let secret_count = lines
        .iter()
        .filter(|line| {
            if let envparser::EnvLine::Assignment { key, value, .. } = line {
                detector::detect(key, value).is_secret
            } else {
                false
            }
        })
        .count();

    // 4. If no secrets, skip protection.
    if secret_count == 0 {
        return Ok(ProtectResult {
            secret_count: 0,
            already_protected: false,
        });
    }

    // 5. Get project hash.
    let hash = vault::project_hash(project_root)
        .map_err(|e| anyhow!("Failed to compute project hash: {e}"))?;

    // Check if already protected (vault already exists).
    let v_path = vault::vault_path(project_root)
        .map_err(|e| anyhow!("Failed to compute vault path: {e}"))?;

    let already_protected = v_path.exists();

    // 6. Generate sandbox content.
    let sandbox_content = sandbox::sandbox_env(&content, &hash);

    // 7. Encrypt original content.
    let vault_bytes =
        vault::encrypt(&content, key).map_err(|e| anyhow!("Vault encryption failed: {e}"))?;

    // 8. Create recovery file bytes.
    let recovery_bytes = recovery::create_recovery_file(key, recovery_key_bytes)
        .map_err(|e| anyhow!("Failed to create recovery file: {e}"))?;

    // 9. Write vault file atomically.
    atomic_write_bytes(&v_path, &vault_bytes)
        .with_context(|| format!("Failed to write vault to {}", v_path.display()))?;

    // 10. Write recovery file atomically.
    let r_path =
        recovery::recovery_path(project_root).context("Failed to compute recovery path")?;
    atomic_write_bytes(&r_path, &recovery_bytes)
        .with_context(|| format!("Failed to write recovery file to {}", r_path.display()))?;

    // 11. Set permissions 600 on vault and recovery files (Unix).
    platform::set_private_permissions(&v_path)
        .with_context(|| format!("Failed to set permissions on {}", v_path.display()))?;
    platform::set_private_permissions(&r_path)
        .with_context(|| format!("Failed to set permissions on {}", r_path.display()))?;

    // 12. Write sandbox content to the .env file on disk (atomic).
    atomic_write_str(&env_path, &sandbox_content)
        .with_context(|| format!("Failed to write sandbox to {}", env_path.display()))?;

    // 13. Update .cloak marker.
    let marker = build_or_update_marker(project_root, rel_path, &hash)?;
    write_marker(project_root, &marker).context("Failed to write .cloak marker")?;

    Ok(ProtectResult {
        secret_count,
        already_protected,
    })
}

/// Remove protection: decrypt vault → restore original `.env` → remove vault →
/// update `.cloak` marker.
///
/// # Errors
///
/// Returns an error if the vault cannot be decrypted, the original content
/// cannot be written to disk, or vault removal fails.
pub fn unprotect_file(project_root: &Path, rel_path: &str, key: &[u8; 32]) -> Result<()> {
    // Decrypt vault and get real content.
    let real_content = read_real(project_root, rel_path, key)?;

    // Restore the real .env to disk.
    let env_path = project_root.join(rel_path);
    atomic_write_str(&env_path, &real_content)
        .with_context(|| format!("Failed to restore {}", env_path.display()))?;

    // Remove vault file.
    let v_path = vault::vault_path(project_root)
        .map_err(|e| anyhow!("Failed to compute vault path: {e}"))?;
    if v_path.exists() {
        std::fs::remove_file(&v_path)
            .with_context(|| format!("Failed to remove vault {}", v_path.display()))?;
    }

    // Update marker — remove this file from the protected list.
    if let Some(mut marker) = read_marker(project_root)? {
        marker.protected.retain(|p| p != rel_path);
        write_marker(project_root, &marker).context("Failed to update .cloak marker")?;
    }

    Ok(())
}

/// Read the real (decrypted) content of a protected `.env` file from its vault.
///
/// # Errors
///
/// - If the vault file is missing, returns an error mentioning `cloak recover`.
/// - If decryption fails (wrong key or corruption), returns an error mentioning `cloak recover`.
/// - If the decrypted content does not contain at least one `=` (sanity check), returns
///   an error mentioning possible corruption.
pub fn read_real(project_root: &Path, rel_path: &str, key: &[u8; 32]) -> Result<String> {
    let v_path = vault::vault_path(project_root)
        .map_err(|e| anyhow!("Failed to compute vault path: {e}"))?;

    // Check for missing vault when a marker is present.
    if !v_path.exists() {
        if read_marker(project_root)?.is_some() {
            return Err(anyhow!(
                "Vault file missing. Run `cloak recover` to restore access."
            ));
        }
        return Err(anyhow!(
            "No vault found for {}. Has this file been protected?",
            rel_path
        ));
    }

    let vault_bytes = std::fs::read(&v_path)
        .with_context(|| format!("Failed to read vault from {}", v_path.display()))?;

    // Quick sanity check before attempting decryption.
    if !vault::is_vault(&vault_bytes) {
        return Err(anyhow!(
            "Vault corrupted. If you have your recovery key, run `cloak recover`."
        ));
    }

    let plaintext = vault::decrypt(&vault_bytes, key)
        .map_err(|e| anyhow!("Wrong key or corrupted vault. Try `cloak recover`. ({})", e))?;

    // Basic integrity check: a valid .env must contain at least one `=`.
    if !plaintext.contains('=') {
        return Err(anyhow!(
            "Vault corrupted: decrypted content does not look like a .env file. \
             If you have your recovery key, run `cloak recover`."
        ));
    }

    Ok(plaintext)
}

/// Save new real content: encrypt to vault and write sandbox version to disk.
///
/// Both the vault and the on-disk `.env` (sandbox) are written atomically.
///
/// # Errors
///
/// Returns an error if encryption fails or any file write fails.
pub fn save_real(project_root: &Path, rel_path: &str, content: &str, key: &[u8; 32]) -> Result<()> {
    // Compute paths.
    let v_path = vault::vault_path(project_root)
        .map_err(|e| anyhow!("Failed to compute vault path: {e}"))?;
    let hash = vault::project_hash(project_root)
        .map_err(|e| anyhow!("Failed to compute project hash: {e}"))?;
    let env_path = project_root.join(rel_path);

    // Encrypt new content.
    let vault_bytes =
        vault::encrypt(content, key).map_err(|e| anyhow!("Vault encryption failed: {e}"))?;

    // Generate sandbox content.
    let sandbox_content = sandbox::sandbox_env(content, &hash);

    // Write vault atomically.
    atomic_write_bytes(&v_path, &vault_bytes)
        .with_context(|| format!("Failed to write vault to {}", v_path.display()))?;

    // Write sandbox to disk atomically.
    atomic_write_str(&env_path, &sandbox_content)
        .with_context(|| format!("Failed to write sandbox to {}", env_path.display()))?;

    Ok(())
}

/// Read the `.cloak` marker file from `<project_root>/.cloak`.
///
/// Returns `Ok(None)` when the file does not exist.
///
/// # Errors
///
/// Returns an error if the file exists but cannot be parsed as valid JSON.
pub fn read_marker(project_root: &Path) -> Result<Option<CloakMarker>> {
    let marker_path = marker_path(project_root);
    if !marker_path.exists() {
        return Ok(None);
    }
    let content = std::fs::read_to_string(&marker_path)
        .with_context(|| format!("Failed to read marker file {}", marker_path.display()))?;
    let marker: CloakMarker = serde_json::from_str(&content)
        .with_context(|| format!("Failed to parse marker file {}", marker_path.display()))?;
    Ok(Some(marker))
}

/// Write (or overwrite) the `.cloak` marker file at `<project_root>/.cloak`.
///
/// The file is written atomically (write-to-`.tmp`, then rename).
///
/// # Errors
///
/// Returns an error if serialization or the file write fails.
pub fn write_marker(project_root: &Path, marker: &CloakMarker) -> Result<()> {
    let marker_path = marker_path(project_root);
    let json = serde_json::to_string_pretty(marker).context("Failed to serialize marker")?;
    atomic_write_str(&marker_path, &json)
        .with_context(|| format!("Failed to write marker to {}", marker_path.display()))?;
    Ok(())
}

// ─────────────────────────────────────────────────────────────────────────────
// Private helpers
// ─────────────────────────────────────────────────────────────────────────────

/// Returns the path to the `.cloak` marker file for the given project root.
fn marker_path(project_root: &Path) -> PathBuf {
    project_root.join(".cloak")
}

/// Returns a simple ISO 8601-style timestamp string (UTC-assumed, no chrono dependency).
///
/// Format: `YYYY-MM-DDTHH:MM:SSZ` derived from [`std::time::SystemTime`].
fn iso8601_now() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};

    let secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);

    // Compute date/time components from Unix timestamp.
    let s = secs % 60;
    let m = (secs / 60) % 60;
    let h = (secs / 3600) % 24;

    // Days since epoch.
    let days = secs / 86400;

    // Gregorian calendar calculation.
    let (year, month, day) = days_to_ymd(days);

    format!("{year:04}-{month:02}-{day:02}T{h:02}:{m:02}:{s:02}Z")
}

/// Convert days since Unix epoch (1970-01-01) to (year, month, day).
fn days_to_ymd(days: u64) -> (u64, u64, u64) {
    // Algorithm from http://howardhinnant.github.io/date_algorithms.html
    let z = days + 719468;
    let era = z / 146097;
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if m <= 2 { y + 1 } else { y };
    (y, m, d)
}

/// Build a new or update an existing [`CloakMarker`], adding `rel_path` to the
/// protected list if not already present.
fn build_or_update_marker(
    project_root: &Path,
    rel_path: &str,
    project_hash: &str,
) -> Result<CloakMarker> {
    let mut marker = match read_marker(project_root)? {
        Some(existing) => existing,
        None => CloakMarker {
            version: version::MARKER_FORMAT_VERSION,
            protected: Vec::new(),
            project_hash: project_hash.to_string(),
            created_at: iso8601_now(),
        },
    };

    if !marker.protected.contains(&rel_path.to_string()) {
        marker.protected.push(rel_path.to_string());
    }

    Ok(marker)
}

/// Atomically write `data` (bytes) to `path`.
///
/// Writes to `<path>.tmp` first, then renames. On Windows the original file
/// is removed before the rename (not atomic, but best-available).
fn atomic_write_bytes(path: &Path, data: &[u8]) -> Result<()> {
    let tmp_path = path.with_extension("tmp");

    std::fs::write(&tmp_path, data)
        .with_context(|| format!("Failed to write temp file {}", tmp_path.display()))?;

    #[cfg(windows)]
    if path.exists() {
        std::fs::remove_file(path)
            .with_context(|| format!("Failed to remove {} before rename", path.display()))?;
    }

    std::fs::rename(&tmp_path, path).with_context(|| {
        format!(
            "Failed to rename {} to {}",
            tmp_path.display(),
            path.display()
        )
    })?;

    Ok(())
}

/// Atomically write `text` (UTF-8) to `path`.
///
/// Convenience wrapper around [`atomic_write_bytes`].
fn atomic_write_str(path: &Path, text: &str) -> Result<()> {
    atomic_write_bytes(path, text.as_bytes())
}
